rasterio/merge.py

```diff
@@ -392,3 +406,354 @@ def nullcontext(obj):
+
+            out_transform: affine.Affine()
+                Information for mapping pixel coordinates in `dest` to another
+                coordinate system
+
+    """
+    if precision is not None:
+        warnings.warn(
+            "The precision parameter is unused, deprecated, and will be removed in 2.0.0.",
+            RasterioDeprecationWarning,
+        )
+
+    if method in MERGE_METHODS:
+        copyto = MERGE_METHODS[method]
+    elif callable(method):
+        copyto = method
+    else:
+        raise ValueError('Unknown method {0}, must be one of {1} or callable'
+                         .format(method, list(MERGE_METHODS.keys())))
+
+    dataset_opener = get_dataset_opener(dataset=datasets[0])
+    with dataset_opener(datasets[0]) as first:
+        first_profile = first.profile
+        first_res = first.res
+        first_crs = first.crs
+        nodataval = first.nodatavals[0]
+        dt = first.dtypes[0]
+
+        if indexes is None:
+            src_count = first.count
+        elif isinstance(indexes, int):
+            src_count = indexes
+        else:
+            src_count = len(indexes)
+
+        try:
+            first_colormap = first.colormap(1)
+        except ValueError:
+            first_colormap = None
+
+    if not output_count:
+        output_count = src_count
+
+    if not crs:
+        crs = first_crs
+    elif type(crs) == str:
+        crs = CRS.from_string(crs)
+
+    # Extent from option or extent of all inputs
+    if bounds:
+        dst_w, dst_s, dst_e, dst_n = bounds
+    else:
+        # scan input files
+        dst_w, dst_s, dst_e, dst_n = get_common_bounds(
+            datasets,
+            crs,
+        )
+
+    # Resolution/pixel size
+    if not res:
+        res = first_res
+    elif not np.iterable(res):
+        res = (res, res)
+    elif len(res) == 1:
+        res = (res[0], res[0])
+
+    if target_aligned_pixels:
+        dst_w = math.floor(dst_w / res[0]) * res[0]
+        dst_e = math.ceil(dst_e / res[0]) * res[0]
+        dst_s = math.floor(dst_s / res[1]) * res[1]
+        dst_n = math.ceil(dst_n / res[1]) * res[1]
+
+    # Compute output array shape. We guarantee it will cover the output
+    # bounds completely
+    output_width = int(round((dst_e - dst_w) / res[0]))
+    output_height = int(round((dst_n - dst_s) / res[1]))
+
+    output_transform = Affine.translation(
+        dst_w, dst_n) * Affine.scale(res[0], -res[1])
+
+    if dtype is not None:
+        dt = dtype
+        logger.debug("Set dtype: %s", dt)
+
+    out_profile = first_profile
+    out_profile.update(**(dst_kwds or {}))
+
+    out_profile["transform"] = output_transform
+    out_profile["height"] = output_height
+    out_profile["width"] = output_width
+    out_profile["count"] = output_count
+    out_profile["dtype"] = dt
+    if nodata is not None:
+        out_profile["nodata"] = nodata
+
+    # create destination array
+    dest = np.zeros((output_count, output_height, output_width), dtype=dt)
```

Comment:
This is an important implementation detail: the merge cannot deal with arbitrarily large datasets that wouldn't fit in RAM. Using [GDALWarp()](https://gdal.org/api/gdal_utils.html#_CPPv48GDALWarpPKc12GDALDatasetHiP12GDALDatasetHPK18GDALWarpAppOptionsPi) would avoid that limitation